src/Processors/QueryPlan/Optimizations/optimizeUseAggregateProjection.cpp

```diff
@@ -249,12 +249,17 @@ static void appendAggregateFunctions(
 
         auto & input = inputs[match.description];
         if (!input)
-            input = &proj_dag.addInput(match.description->column_name, std::move(type));
+            input = &proj_dag.addInput(match.description->column_name, type);
 
         const auto * node = input;
 
         if (node->result_name != aggregate.column_name)
-            node = &proj_dag.addAlias(*node, aggregate.column_name);
+        {
+            /// Always cast to aggregate types specified in query, because input
+            /// columns from projection might have the same state but different
+            /// type, which can generate wrong results during finalization.
+            node = &proj_dag.addCast(*node, type, aggregate.column_name);
```

Comment:
It is almost ok, but can we check that the types are different separately?
I think it may be possible that types are equals, but names are different.
Can we add `cast` only if types are different, and `alias` only if names are different?

Reply:
Hmm, I don't think it's possible after #39420 , otherwise we need a different `equals` method for DataTypeAggregateFunction